src/pydvl/utils/parallel/futures/ray.py

```diff
@@ -323,13 +347,18 @@ def flag_executor_shutting_down(self):
             logger.debug("work item manager thread acquired shutdown lock")
             if executor is not None:
                 executor._shutdown = True
-                # Cancel pending work items if requested.
-                if executor._cancel_pending_futures:
+                if executor._cancel_futures & CancellationPolicy.RUNNING:
                     logger.debug("forcefully cancelling running futures")
                     # We cancel the future's object references
                     # We cannot cancel a running future object.
                     for future in self.submitted_futures:
                         ray.cancel(future.object_ref)
+                    # FIXME: huh? why is this here?
+                    # Make sure we do this only once to not waste time looping
+                    # on running processes over and over.
+                    executor._cancel_futures = CancellationPolicy.NONE
+
+                if executor._cancel_futures & CancellationPolicy.PENDING:
```

Comment:
I think this block should run first to drain the pending and work queues to prevent submitting any further jobs and then we can cancel the running.